components/merkledb/src/proof_list_index/proof.rs

```diff
@@ -58,17 +58,77 @@ impl ProofOfAbsence {
 
 /// An enum that represents a proof of existence for a proof list elements.
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub enum ListProof<V> {
+pub enum ProofVariant<V> {
     /// A branch of proof in which both children contain requested elements.
-    Full(Box<ListProof<V>>, Box<ListProof<V>>),
+    Full(Box<ProofVariant<V>>, Box<ProofVariant<V>>),
     /// A branch of proof in which only the left child contains requested elements.
-    Left(Box<ListProof<V>>, Option<Hash>),
+    Left(Box<ProofVariant<V>>, Option<Hash>),
     /// A branch of proof in which only the right child contains requested elements.
-    Right(Hash, Box<ListProof<V>>),
+    Right(Hash, Box<ProofVariant<V>>),
     /// A leaf of the proof with the requested element.
     Leaf(V),
     /// Proof of absence of an element with the specified index.
-    Absent(ProofOfAbsence),
+    Absent(Hash),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ListProof<V> {
+    length: u64,
+    proof: ProofVariant<V>,
+}
+
+impl<V> ListProof<V>
+where
+    V: BinaryValue,
+{
+    pub(crate) fn new(length: u64, proof: ProofVariant<V>) -> Self {
+        Self { length, proof }
+    }
+
+    /// Length of the corresponding ProofList.
+    pub fn length(&self) -> u64 {
+        self.length
+    }
+
+    /// Underlying proof variant.
+    pub fn proof_variant(&self) -> &ProofVariant<V> {
+        &self.proof
+    }
+
+    /// Verifies the correctness of the proof by the trusted list hash and the number of
```

Comment:
'the number of elements in the tree' is no longer true, as this method accepts the expected list hash only.